app/server/lib/AttachmentFileManager.ts

```diff
@@ -108,24 +118,258 @@ export class AttachmentFileManager implements IAttachmentFileManager {
+  public startTransferringFileToOtherStore(fileIdent: string, newStoreId: AttachmentStoreId | undefined) {
+    this._pendingFileTransfers.set(fileIdent, newStoreId);
+    this._runTransferJob();
+  }
+
+  // Generally avoid calling this directly, instead use other methods to schedule and run the
+  // transfer job. If a file with a matching identifier already exists in the new store, no
+  // transfer will happen, as the default _addFileToX behaviour is to avoid re-uploading files.
+  public async transferFileToOtherStore(fileIdent: string, newStoreId: AttachmentStoreId | undefined): Promise<void> {
+    this._log.info({fileIdent, storeId: newStoreId}, `transferring file to new store`);
+    const fileMetadata = await this._docStorage.getFileInfoNoData(fileIdent);
+    // This check runs before the file is retrieved as an optimisation to avoid loading files into
+    // memory unnecessarily.
+    if (!fileMetadata || (fileMetadata.storageId ?? undefined) === newStoreId) {
+      return;
+    }
+    // It's possible that the record has changed between the original metadata check and here.
+    // However, the worst case is we transfer a file that's already been transferred, so no need to
+    // re-check.
+    // Streaming isn't an option here, as SQLite only supports buffers (meaning we need to keep at
+    // least 1 full copy of the file in memory during transfers).
+    const file = await this._getFile(fileIdent);
+    if (!await validateFileChecksum(fileIdent, file.data)) {
+      throw new AttachmentRetrievalError(file.storageId, file.ident, "checksum verification failed for retrieved file");
+    }
+    if (!newStoreId) {
+      await this._storeFileInLocalStorage(fileIdent, file.data);
+      return;
+    }
+    const newStore = await this._getStore(newStoreId);
+    if (!newStore) {
+      this._log.warn({
+        fileIdent,
+        storeId: newStoreId
+      }, `unable to transfer file to unavailable store`);
+      throw new StoreNotAvailableError(newStoreId);
+    }
+    // Store should error if the upload fails in any way.
+    await this._storeFileInAttachmentStore(newStore, fileIdent, file.data);
+
+    // Don't remove the file from the previous store, in case we need to roll back to an earlier
+    // snapshot.
+    // Internal storage is the exception (and is automatically erased), as that's included in
+    // snapshots.
+  }
+
+  public async allTransfersCompleted(): Promise<void> {
+    if (this._transferJob) {
+      await this._transferJob.promise;
+    }
+  }
+
+  public transferStatus() {
+    return {
+      pendingTransferCount: this._pendingFileTransfers.size,
+      isRunning: this._transferJob !== undefined
+    };
+  }
+
+  private _runTransferJob(): TransferJob {
+    if (this._transferJob) {
+      return this._transferJob;
+    }
+    const transferPromise = this._performPendingTransfers();
+    const newTransferJob: TransferJob = {
+      promise: transferPromise,
```

Comment:
Why the nesting? Could newTransferJob not be assigned transferPromise directly?

Reply:
I think I was anticipating adding more info to this later - but I can simplify for now and deal with that if / when we get to it.